### PR TITLE
Rebuild website to use an Angular UI.

### PR DESCRIPTION
**Pull request by tdsymonds**
The rebuild resolves an issue with Consent Mode and adds new functionality:
- There's now an event bar that shows you the e-commerce events, so you don't have to inspect the code.
- There are product variants, so you can buy different sizes/colours of the products

gtm-boilerplate-demo1
gtm-boilerplate-demo2
gtm-boilerplate-demo3

### CODE REVIEW

File: website/ui/src/app/services/ecommerce-events.service.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+      item_variant: productVariant.name,
+    };
+  }
+
+  /**
+   * Output the event to the console.
+   * @param event the event to output
+   */
+  private logEvent(event: EcommerceEvent): void {
+    console.log('Generated ecommerce event:');
```
Comment (HalleyInteractive) [inline website/ui/src/app/services/ecommerce-events.service.ts:113]: You can also do `console.log('Generated ecommerce event: ', event);`

File: website/ui/src/app/services/consent.service.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    this.updateGtagConsent(ConsentUpdate.DEFAULT);
+    this.getConsentFromCookie();
+    if (this.hasConsentCookie === true) {
+      this.updateGtagConsent(ConsentUpdate.UPDATE);
+    }
+    this.isInitialized.next(true);
+  }
+
+  /**
+   * Get the current consent
```
Comment (HalleyInteractive) [inline website/ui/src/app/services/consent.service.ts:76]: Add `.`

File: website/ui/src/app/services/basket.service.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  constructor(
+    private cookieService: CookieService,
+    private router: Router,
+    private productsService: ProductsService,
+  ) {
+    this.refreshBasketFromCookie();
+  }
+
+  /**
+   * Get the basket
```
Comment (HalleyInteractive) [inline website/ui/src/app/services/basket.service.ts:54]: Add `.`

**Review by HalleyInteractive — APPROVED** [review]
Hey Tom, really nice update to the GTM Boilerplate. Thanks for adding this. 
I've left a couple of small nits only.

File: website/ui/src/app/components/login-form/login-form.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  describe('ngOnInit()', () => {
+    it('should initialize form with user data from LoginService', () => {
+      const mockUserData = {
```
[the remark this answers is not in the document]
Comment (tdsymonds) [reply website/ui/src/app/components/login-form/login-form.component.spec.ts:62]: Nope. That's a mistake. Good spot, thanks.

File: website/ui/src/app/components/login-form/login-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  constructor(public loginService: LoginService) {}
+
+  ngOnInit(): void {
+    // Initialize form with user data (if available)
+    const u = this.loginService.user;
+    const uid = u.id;
+    const uname = u.name;
+    console.log(u);
+    console.log(u.id);
```
[the remark this answers is not in the document]
Comment (tdsymonds) [reply website/ui/src/app/components/login-form/login-form.component.ts:43]: Oops, those were from some ugly troubleshooting. Removed. Thanks

File: website/ui/src/app/services/ecommerce-events.service.ts [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    this.gtmService.pushTag(event);
+    this.events.unshift(this.formatEcommerceEventAsString(event));
+  }
+
+  /**
+   * Send the purchase event to GTM.
+   * @param basket the basket containing the products.
+   * @param value the total value of the basket.
+   * @param transaction_id the ID of the transaction.
+   * @return a purchase ecommerce event.
```
[the remark this answers is not in the document]
Comment (tdsymonds) [reply website/ui/src/app/services/ecommerce-events.service.ts:332]: Doh! Copy and paste working against me.

File: website/ui/src/app/services/ecommerce-events.service.ts [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+    );
+    this.gtmService.pushTag(event);
+    this.events.unshift(this.formatEcommerceEventAsString(event));
+  }
+
+  /**
+   * Send the view_cart event to GTM.
+   * @param basket the basket containing the products.
+   * @param value the total value of the basket.
+   * @return a view_cart ecommerce event.
```
[the remark this answers is not in the document]
Comment (tdsymonds) [reply website/ui/src/app/services/ecommerce-events.service.ts:317]: Ditto

File: website/ui/src/app/services/ecommerce-events.service.ts (hunk 1 above)
Comment (tdsymonds) [reply website/ui/src/app/services/ecommerce-events.service.ts:113]: Good callout. Done.

File: website/ui/src/app/app.component.html [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+-->
+<app-top-bar></app-top-bar>
+<div class="main-container">
+  <div class="container-fluid">
+    <div class="row">
+      <div class="sidebar" [ngClass]="{'collapsed': !showSidebar, 'col-md-auto': showSidebar}">
+        <img class="hamburger-icon"
+            src="assets/images/hamburger-icon.png"
+            (click)="toggleSidebar()"/>
+        <div [ngClass]="{'hide': !showSidebar}">
```
[the remark this answers is not in the document]
Comment (tdsymonds) [reply website/ui/src/app/app.component.html:24]: Good point. Updated. Thanks

File: website/ui/src/app/services/basket.service.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+  /**
+   * Get the basket contents from the cookie.
+   * @return the basket.
+   */
+  private getBasketFromCookie(): Basket {
+    const basketJson = this.cookieService.get(this.cookieName);
+    if (basketJson == null || basketJson === '') {
+      return {};
+    } else {
+      return this.parseCookieToBasket(JSON.parse(basketJson));
```
[the remark this answers is not in the document]
Comment (tdsymonds) [reply website/ui/src/app/services/basket.service.ts]: Good shout. I've implemented that logic now.

**Comment by tdsymonds** [comment]
@HalleyInteractive I've made some changes based on the feedback you provided. Please re-review and let me know if anything further is needed. Thanks

**Review by HalleyInteractive — APPROVED** [review]

**Pull request merged**